### PR TITLE
[BugFix] Propagate 'trust_remote_code' setting in internvl and minicpmv

### PR DESCRIPTION
**Pull request by zifeitong**
For MiniCPM-V model, the token ids needed for compute image bounds are piped through the preprocessing pipeline via `MiniCPMVImageInput` and `MultiModalInputs`.

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/models/minicpmv.py [hunk 1]
```diff
@@ -400,13 +405,11 @@ def get_embedding(
         return vlm_embedding, vision_hidden_states
 
     def _get_image_bounds(self, input_ids: torch.Tensor) -> torch.Tensor:
-        tokenizer = cached_get_tokenizer(self.config._name_or_path,
-                                         trust_remote_code=True)
-        start_cond = input_ids == tokenizer.im_start_id
-        end_cond = input_ids == tokenizer.im_end_id
-        if hasattr(tokenizer, "slice_start_id"):
-            start_cond |= (input_ids == tokenizer.slice_start_id)
-            end_cond |= (input_ids == tokenizer.slice_end_id)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/minicpmv.py:409]: Can we pass this information via the multi-modal kwargs?

**Review by DarkLight1337 — COMMENTED** [review]
The change to InternVL looks good to me. As for MiniCPM-V:

File: vllm/model_executor/models/minicpmv.py (hunk 1 above)
Comment (zifeitong) [reply vllm/model_executor/models/minicpmv.py:409]: You mean MultiModalConfig?

Comment (DarkLight1337) [reply vllm/model_executor/models/minicpmv.py:409]: I mean that apart from the image input data (e.g. `pixel_values`), we may be able to return these IDs via the input processor which does have access to the tokenizer.

Comment (DarkLight1337) [reply vllm/model_executor/models/minicpmv.py:409]: Nice work, this is much cleaner now!

**Review by DarkLight1337 — APPROVED** [review]
Thanks for addressing my comments, this looks good now!

**Comment by zifeitong** [comment]
Sorry for the delay. I still need to get the tests passed. I'll let you know once it's working.

**Comment by zifeitong** [comment]
@DarkLight1337 PTAL. I have to register another plugin for image_bounds aux data.

**Comment by DarkLight1337** [comment]
Rather than using another plugin, I suggest that you override the default input mapper (see Fuyu model for an example).

**Comment by zifeitong** [comment]
> Rather than using another plugin, I suggest that you override the default input mapper (see Fuyu model for an example).

I tried that idea, but input mapper can't access token ids. Am I missing anything?

**Comment by DarkLight1337** [comment]
I mean that you can override the input mapper so that it accepts the image bounds without having to define another plugin to process it.

**Comment by zifeitong** [comment]
> I mean that you can override the input mapper so that it accepts the image bounds from the input processor without having to define another plugin to handle it.

`map_input` is tied to plugins:
https://github.com/vllm-project/vllm/blob/3118f63385c0d767fba8b6d2039fc35440678da9/vllm/multimodal/registry.py#L116

I would get a `Unknown multi-modal data type: image_bounds` error if i didn't register a plugin.

**Comment by DarkLight1337** [comment]
Oh, I forgot about that part... perhaps you can change the type of data passed to `"image"` key instead. For example,

```py
multi_modal_data["image"] = {"data:" image, "bounds": image_bounds}
```

and override the input mapper to handle this case.

**Comment by zifeitong** [comment]
> multi_modal_data["image"] = {"data": image, "bounds": image_bounds}

I tried it out and it's also very awkward:

`multi_modal_data["image"] = {"data": List[Image], "bounds": image_bounds}` won't work with `--limit-mm-per-prompt`, since the image input is no longer a list.

On the other hand, `multi_modal_data["image"] = [{"data": List[Image], "bound": image_bound}]` also won't work, since `len(image_bounds)` is not always the same as `len(images)`. `image_bounds` is computed completely on prompt tokens.

**Comment by DarkLight1337** [comment]
> On the other hand, `multi_modal_data["image"] = [{"data": List[Image], "bound": image_bound}]` also won't work, since `len(image_bounds)` is not always the same as `len(images)`. `image_bounds` is computed completely on prompt tokens.

If the number of images represented by the prompt isn't the same as the number of input images, it is probably a user error. IMO, an error should be thrown in that case anyway.

**Comment by zifeitong** [comment]
> > On the other hand, `multi_modal_data["image"] = [{"data": List[Image], "bound": image_bound}]` also won't work, since `len(image_bounds)` is not always the same as `len(images)`. `image_bounds` is computed completely on prompt tokens.
> 
> If the number of images represented by the prompt isn't the same as the number of input images, it is probably a user error. IMO, an error should be thrown in that case anyway.

It's related to the "slice" concept used in the model which I am not familiar with https://github.com/vllm-project/vllm/blob/a91165f7afe08ad47750c6d5270471aa0242e27f/vllm/model_executor/models/minicpmv.py#L298. The length of image_bounds can be much larger than number of images.

How's the latest version with input mapper? "image_bounds" is carried around with every input image which is less than ideal. But otherwise pretty clean to me.

File: vllm/model_executor/models/minicpmv.py [hunk 2]
```diff
@@ -69,6 +70,15 @@
 }
 
 
+class MiniCPMVImageInput(TypedDict):
+    """Image input to MiniCPM-V model with image bound auxiliary data."""
+
+    image: Image.Image
+
+    # If provided, contains image_bounds data for the whole prompt.
+    image_bounds: Optional[torch.Tensor]
```
Comment (DarkLight1337) [inline vllm/model_executor/models/minicpmv.py:79]:
```suggestion
    image_bounds: NotRequired[torch.Tensor]
```

Should use `NotRequired` from `typing_extensions`.

File: vllm/model_executor/models/minicpmv.py [hunk 3]
```diff
@@ -331,6 +371,30 @@ def get_placeholder(image_size: Tuple[int, int], num_image: int):
...
+    if image_processor is None:
+        raise RuntimeError("No HuggingFace processor is available "
+                           "to process the image object")
+    try:
+        batch_data = image_processor \
+            .preprocess([img["image"] for img in data], return_tensors="pt") \
+            .data
+    except Exception:
+        logger.error("Failed to process image (%s)", data)
+        raise
```
Comment (DarkLight1337) [inline vllm/model_executor/models/minicpmv.py:388]: Assert that `data` is a list?

File: vllm/model_executor/models/minicpmv.py [hunk 4]
```diff
@@ -331,6 +371,30 @@ def get_placeholder(image_size: Tuple[int, int], num_image: int):
...
+
+    for image in data:
+        if "image_bounds" in image:
+            return MultiModalInputs(image_bounds=image["image_bounds"],
+                                    **batch_data)
+
+    return MultiModalInputs(batch_data)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/minicpmv.py:395]:
```suggestion
    if len(data) > 0:
        batch_data["image_bounds"] = data[0]["image_bounds"]

    return MultiModalInputs(batch_data)
```

File: vllm/model_executor/models/minicpmv.py [hunk 5]
```diff
@@ -387,7 +451,7 @@ def get_embedding(
             if len(image_bounds) > 0:
                 image_indices = torch.stack([
                     torch.arange(start, end, dtype=torch.long)
-                    for start, end in image_bounds.tolist()
+                    for start, end in image_bounds.squeeze(0).tolist()
```
Comment (DarkLight1337) [inline vllm/model_executor/models/minicpmv.py:454]: To keep things consistent, let's flatten `image_bounds` inside `_parse_and_validate_input`.

**Review by DarkLight1337 — COMMENTED** [review]
Yeah, this looks reasonable. A few more comments.

**Comment by zifeitong** [comment]
PTAL. Comments addressed.

File: vllm/model_executor/models/minicpmv.py [hunk 6]
```diff
@@ -379,6 +380,10 @@ def input_mapper_for_minicpmv(ctx: InputContext, data: object):
     if image_processor is None:
         raise RuntimeError("No HuggingFace processor is available "
                            "to process the image object")
+
+    if not isinstance(data, list) or len(data) <= 0:
+        raise ValueError("Invalid image input (%s)", data)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/minicpmv.py:385]: Users are currently allowed to pass an empty list of images. We should still allow this case.

File: vllm/model_executor/models/minicpmv.py [hunk 7]
```diff
@@ -482,6 +487,11 @@ def _parse_and_validate_inputs(
             raise ValueError("Inconsistent batch lengths, found: "
                              f"{len(pixel_values)} vs. {len(tgt_sizes)}")
 
+        if not isinstance(image_bounds, torch.Tensor):
+            raise ValueError("Incorrect type of image_bounds. "
+                             f"Got type: {type(image_bounds)}")
+        image_bounds = image_bounds.squeeze(0)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/minicpmv.py:493]: Why `squeeze`? From my understanding, this may not hold true for multiple items in a batch.

Comment (zifeitong) [reply vllm/model_executor/models/minicpmv.py:493]: The `image_bounds` tensor is stacked here: [    def _try_stack(nested_tensors: NestedTensors) -> NestedTensors:](https://github.com/vllm-project/vllm/blob/18ae428a0d8792d160d811a9cd5bb004d68ea8bd/vllm/multimodal/base.py#L49)

Note that L54 will never be triggered since the input to try_stack will always be List.

Comment (DarkLight1337) [reply vllm/model_executor/models/minicpmv.py:493]: I understand that part. But `_try_stack` will stack the tensor to the batch size, which may be greater than 1. In that case you can't rely on `squeeze` anymore.

Comment (zifeitong) [reply vllm/model_executor/models/minicpmv.py:493]: > I understand that part. But _try_stack will stack the tensor to the batch size, which may be greater than 1

you're right. i can't see how the current code can work when batch size is greater than 1. let me check and maybe add some test cases.

Comment (zifeitong) [reply vllm/model_executor/models/minicpmv.py:493]: The current code works since `_get_image_bounds()` uses the batch (concatenated) token_ids as input.

I'll change the code to pass `im_start_id` and `im_end_id` as auxiliary data, so we can again run `_get_image_bounds()` on batched token_ids (instead of at per request level in input processor).

Comment (DarkLight1337) [reply vllm/model_executor/models/minicpmv.py:493]: Sounds good!

Comment (zifeitong) [reply vllm/model_executor/models/minicpmv.py]: PTAL.

Batch size > 1 works fine now.

**Comment by zifeitong** [comment]
> Can you run `format.sh` to fix the lint errors?

Done. Sorry about that.

File: vllm/model_executor/models/minicpmv.py [hunk 8]
```diff
@@ -393,14 +457,18 @@ def get_embedding(
...
+            input_ids: torch.Tensor,
+            im_start_id: torch.Tensor,
+            im_end_id: torch.Tensor,
+            slice_start_id: Optional[torch.Tensor] = None,
+            slice_end_id: Optional[torch.Tensor] = None) -> torch.Tensor:
+        start_cond = input_ids == im_start_id[0]
+        end_cond = input_ids == im_end_id[0]
+        if slice_start_id is not None:
+            start_cond |= (input_ids == slice_start_id[0])
+            end_cond |= (input_ids == slice_end_id[0])
```
Comment (DarkLight1337) [inline vllm/model_executor/models/minicpmv.py:471]: Can you add a note here that we assume these token IDs to be the same across batches?

**Review by DarkLight1337 — APPROVED** [review]

**Comment by zifeitong** [comment]
Thanks a lot for the review!
